libqtile/group.py

```diff
@@ -90,16 +93,33 @@ def current_window(self, win):
             pass
         self.focus_history.append(win)
 
-    def _remove_from_focus_history(self, win):
-        try:
-            index = self.focus_history.index(win)
-        except ValueError:
-            # win has never received focus
-            return False
-        else:
-            del self.focus_history[index]
-            # return True if win was the last item (i.e. it was current_window)
-            return index == len(self.focus_history)
+    def _order_clients(self):
+        """
+        Orders our clients by their layouts z-value, then by the clients z-values.
+        If z-values are equal, the window ids are used for reproducibility.
+        The z-values themselves are managed by the layouts.
+        """
+        return sorted(
+            [w for w in self.windows if not w.minimized],
+            key=lambda c: (c.floating, c.z, c.window.wid)
+        )
+
+    def assure_correct_layer(self, win):
+        """
+        Make sure that the window is in its designated z-position by moving other
+        windows around as necessary.
+        """
+        if win is None or win.minimized:
+            return
+
+        windows = self._order_clients()
+        if self._previous_layering and self._previous_layering == windows:
+            return
+        self._previous_layering = windows
+
+        index = windows.index(win)
+        for w in windows[index:]:
+            w.window.configure(stackmode=StackMode.Above)
```

Comment:
Does this reduce calls? (untested, may not actually make sense):
```
index = windows.index(win)
if index > len(windows) // 2:
    for w in windows[index:]:
        w.window.configure(stackmode=StackMode.Above)
else:
    for w in windows[:index]:
        w.window.configure(stackmode=StackMode.Below)
```

Reply:
That's actually what I initially tried to do, back when the z-updates where tied to changing the `current_window`. In there, however, moving the current window down led to focus changes and thus to an infinite recursion. I'll check how things look now.

Reply:
Other than some trouble with my picture-in-pciture fake screen (which never really worked anyway, which is how this entire z-business started), it seems valid to do this now. At least I didn't observe any problems. I'll put this in the next commit.